File: challenge #1 - Hacker language/typescript/main.ts
import * as fs from 'fs';
import * as path from 'path';

const translateTextFunc = (text: string): string => {
  const leetDictionary: Record<string, string> = {
    'A': '4',
    'B': '|3',
    'C': '[',
    'D': ')',
    'E': '3',
    'F': '|=',
    'G': '&',
    'H': '#',
    'I': '1',
    'J': ',_|',
    'K': '>|',
    'L': '|_',
    'M': '/\\/\\',
    'N': '^/',
    'O': '0',
    'P': '|*',
    'Q': '(_,)',
    'R': '|2',
    'S': '5',
    'T': '7',
    'U': '(_)',
    'V': '\\/',
    'W': '\\/\\/',
    'X': '><',
    'Y': '`/',
    'Z': '2',
    '1': 'L',
    '2': 'R',
    '3': 'E',
    '4': 'A',
    '5': 'S',
    '6': 'b',
    '7': 'T',
    '8': 'B',
    '9': 'g',
    '0': '()',
  }

  const textTranslated = Array
    .from((text), (character) => leetDictionary[character.toUpperCase()] || character)
    .join('');

  return textTranslated;
}

const getInputText = (): Promise<string> => new Promise((resolve, reject) => {
  const chunks: Buffer[] = [];
  const reader = fs.createReadStream(path.join('..', 'text.txt'));

  reader.on('data', (chunk: Buffer) => chunks.push(chunk));

  reader.on('error', (error) => reject(error));

  reader.on('end', () => {
    const inputText = Buffer.concat(chunks).toString();
    resolve(inputText);
  });
});

const main = async (): Promise<void> => {
  const inputText: string = await getInputText();

  const executionRate: number = 1_000;

  const executionTime: number = Array
    .from({ length: executionRate }, () => {
      const startTime = performance.now();
      translateTextFunc(inputText);
      const endTime = performance.now();
      return endTime - startTime;
    })
    .reduce((prev, curr) => prev + curr, 0);

  const startTime = performance.now();
  const textTranslated = translateTextFunc(inputText);
  const endTime = performance.now();

  console.log(`** result: ${textTranslated}\n** function execution time: ${endTime - startTime}\n** average function execution time: ${executionTime / executionRate}`);
}

main();

File: challenge #1 - Hacker language/python/main.py
import functools
import time
import timeit


def translateText(inputText):
  leet_keys = [
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z',
  '1',
  '2',
  '3',
  '4',
  '5',
  '6',
  '7',
  '8',
  '9',
  '0',
  ]

  leet_values = [
  '4',
  '|3',
  '[',
  ')',
  '3',
  '|=',
  '&',
  '#',
  '1',
  ',_|',
  '>|',
  '|_',
  '/\\/\\',
  '^/',
  '0',
  '|*',
  '(_,)',
  '|2',
  '5',
  '7',
  '(_)',
  '\\/',
  '\\/\\/',
  '><',
  '`/',
  '2',
  'L',
  'R',
  'E',
  'A',
  'S',
  'b',
  'T',
  'B',
  'g',
  '()',
  ]

  leet_dictionary = dict(zip(leet_keys, leet_values))
  translated = map(lambda character: leet_dictionary.get(character.upper(), character), inputText)
  text_translated_list = list(translated)
  text_translated = ''.join(text_translated_list)
  return text_translated

start_time = time.time()
text_translated = translateText("Hello world, 123")
end_time = time.time()

function_timeit = timeit.timeit(functools.partial(translateText, "Hello world, 123"), number=1000)

print(f"** function execution time: {end_time - start_time}\n** function average execution time: {function_timeit}\n** result: {text_translated}")
